Schemas.md

```diff
@@ -0,0 +1,521 @@
+
+Oppia uses lots of forms, and it's tiresome to have to write each one from scratch each time. To get around this, we have built a schema-based form framework that allows such forms to be built declaratively. From each schema, we know the type of the form field, what restrictions we want to impose on user input (from the validators), and how the form should be displayed (from the `ui_config` key).
+
+Note that this framework isn't meant to be exhaustive; it is meant to simplify form creation for the more common cases. Custom forms can still be written using the methods referenced in [[Creating Objects|Creating-Objects]].
+
+In the frontend, the `components/forms/schema-based-editors/` directory contains the `schema-based-editor` component which takes a schema and generates a form whose return value satisfies the schema. The schema can be annotated with UI configuration options to tell the form how to display itself. You can also use FormBuilder to create a custom form; the only requirement is that the return type matches that defined by the schema.
+
+For example, the [`story-node-editor` component](https://github.com/oppia/oppia/blob/develop/core/templates/pages/story-editor-page/editor-tab) includes a schema-based editor with the following HTML:
+
+```html
+<schema-based-editor id="storyNodeOutline"
+                     schema="OUTLINE_SCHEMA"
+                     local-value="editableOutline">
+</schema-based-editor>
+```
+
+The `OUTLINE_SCHEMA` constant in the component's TypeScript file like this:
+
+```ts
+$scope.OUTLINE_SCHEMA = {
+  type: 'html',
+  ui_config: {
+    startupFocusEnabled: false,
+    rows: 100
+  }
+};
+```
+
+The result is a form where a user can provide an outline with some basic HTML markup:
+
+![Screenshot of a chapter outline editor](images/schemaBasedEditor.png)
+
+### Validating controller arguments
+
+All arguments passed to the GET/POST/PUT/DELETE methods of the handler classes in the Oppia controller layer or in main.py need to be robustly validated before being passed to the domain layer in the backend. If we define schemas for the expected arguments, then we can use a generic schema validation system (SVS) to validate the arguments!
+
+#### Schema validation system code
+
+The following key methods are used in the validation of handler args through the SVS architecture:
+
+* `vaidate_and_normalize_args()` in `base.py`: This method is defined in the BaseHandler class of base.py. The `vaidate_and_normalize_args` method is responsible for raising all kinds of errors in the context of validation of handler args, like `InvalidInputException` and `NotImplemented` error. (See [this section](#common-errors-faced) for a list of common errors that may arise.)
+
+* `validate(handler_args, handler_args_schemas)` in `payload_validator.py`: This method is the core method for SVS functionality. It collects all the AssertionErrors raised from schema_utils.
+  * `handler_args`: The arguments from payload/ request.
```

Comment:
Maybe elaborate on this a bit more, I don't know what `payload/ request` is.

Reply:
I changed this to `HTTP request` to be clearer